Add dev doc for CI test stage (#2382)
scality/metalk8s: docs/developer/development/ci.rst

Continuous Testing
==================

Add a new test in the continuous integration system
---------------------------------------------------

A test as continuous integration system level is an End-to-End test that need a

> dedicated environment (with one or several machines (virtual or container)).

Reviewer: No? What about unit tests, build/linting steps?

If the goal of this document is to only describe e2e tests and how to develop them for the CI, maybe we need to rephrase this:

```suggestion
In this context, we consider only End-to-End tests that need a dedicated
environment (with one or several machines, being VMs or containers).
```

Author: Check https://github.com/scality/metalk8s/pull/2382#discussion_r406295732

> A test as continuous integration system level is an End-to-End test that need a
> dedicated environment (with one or several machines (virtual or container)).

Reviewer 2: Not sure for this one:
```suggestion
When we refer to test, at continuous integration system level, it means an end-to-end test
that requires a dedicated environment, with one or several machines (virtual or container).
```

Reviewer 3: If i could propose then =>
`At a continuous integration system level, a test is an end-to-end (operation or task) that requires a dedicated environment with one or several machines(virtual or container)`

Author: Done, a mix of both (I think `task` is good and also cover @gdemonet comment https://github.com/scality/metalk8s/pull/2382#discussion_r406150647

A test that only checks a specific feature of a classic MetalK8s deployment
should be part of PyTest BDD and not integrated as a dedicated stage in

> continuous integration system.

Reviewer: This message has a lot of value, but I'm not sure about the description "_only checks a specific feature of a classic MetalK8s deployment_"... What does it mean in practice?

Would be interesting to make it a section for describing the "test implementation strategies" - maybe for a follow-up ticket though.

Author: Not sure we want to talk about it in this CI document, maybe I can just add an example here. Like `testing  Ingress Pod Ready`

How to choose between Pre-merge and Post-merge
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The choice really depends on the goals of this test.

As a high-level view:

Pre-merge:

- Test is usually not long and could last less than 30 minutes.
- Test essential features of the product (installation, expansion, building,
  ...).

Post-merge:

- Test last longer (more than 30 minutes).

> - Test "non-essential" feature of the product (upgrade, downgrade, solutions,

Reviewer 3: I am a little bit sceptical about `non-essential` in this sentence. Maybe we leave it at:
`Test the following features of the product (upgrade, downgrade, solutions ...).`

Author: Yes but the list is not exhaustif so .... 
Maybe I can just add `(not mandatory to have a working cluster)`

  ...).

How to add a stage in continuous integration system

> ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Reviewer: Could be worth adding a paragraph to explain how to make some stages optional (using build properties and `doStepIf`), and also remind that it's possible to group related stages under a parent stage, using the `TriggerStages` method.

Author: Hmm for me it's part of `eve` documentation, we shouldn't talk about it here (even if I know that opensource world do not have access to eve documentation, I think they will not develop stuff for eve because they do not have any access to eve, build, artifacts, ....)

Reviewer: OK. That was more to provide guidance to our own devs, but I get why you'd rather have this in Eve docs directly.

Continuous integration system is controlled by the ``eve/main.yml`` YAML file.

A stage is defined by a worker and a list of steps. Each stage should be in
the ``stages`` section and triggered by ``pre-merge`` or ``post-merge``.

To know the different kind of workers available, all the builtin steps, how to

> trigger a stage, ... please refer to the eve documentation.

Reviewer 2: Could be cool to add a link to the documentation https://eve.devsca.com/doc/

Reviewer 3: Good point @alexandre-allard-scality  even though I think this link is only accessible internally but this doc will be visible to the outside world. One could question this addition?

Author: Yes, I think we shouldn't as it's not opensource

Reviewer 2: Ok, good to me.

A test stage in MetalK8s context
""""""""""""""""""""""""""""""""

> In MetalK8s context each test stage (eve stage that represent a full test)

Reviewer 4: ```suggestion
In MetalK8s context each test stage (eve stage that represents a full test)
```

should generate a status file containing the result of the test, either a
success or a failure, and a JUnit file containing the result of the test
and information about this test.

To generate the JUnit file, each stage needs the following information:

> - The name of the Test Suite this test stage is part of.

Reviewer 2: No need for a period at the end of the line here

> - Section path to regroup test in a Test Suite if needed (optional)

Reviewer 2: ```suggestion
- Section path to group tests in a Test Suite, if needed (optional)
```

- A test name

Before executing all the steps of the test we first generate a failed
result and at the end of the test we generate a success result. So that the

> failed result get overrided by the success one if everything goes well.

Reviewer 2: ```suggestion
failed result get overridden by the success one if everything goes well.
```

At the very end, the final status of a test should be uploaded no matter the
outcome of the test.

> To generate these result we already have several helpers available.

Reviewer 2: ```suggestion
To generate these results, we already have several helpers available.
```

Example:

Consider we want a new test named ``My Test`` which is part of
the subsection ``My sub section`` of the section ``My section`` in the
test suite ``My Test Suite``.

.. note::

>   Test, suite, class names are not case sensitive in ``eve/main.yml``.

Reviewer 2: ```suggestion
  Test, suite and class names are not case sensitive in ``eve/main.yml``.
```

.. code-block:: yaml

  my-stage:
    _metalk8s_internal_info:
      junit_info: &_my_stage_junit_info
        TEST_SUITE: my test suite
        CLASS_NAME: my section.my sub section
        TEST_NAME: my test
    worker:
      # ...
      # Worker informations
      # ...
    steps:
      - Git: *git_pull
      - ShellCommand: # Generate a failed final status
          <<: *add_final_status_artifact_failed
          env:
            <<: *_env_final_status_artifact_failed
            <<: *_my_stage_junit_info
            STEP_NAME: my-stage
       # ...
       # All test steps should be here !
       # ...
       - ShellCommand: # Generate a success final status
         <<: *add_final_status_artifact_success
         env:
           <<: *_env_final_status_artifact_success
           <<: *_my_stage_junit_info
           STEP_NAME: my-stage
       - Upload: *upload_final_status_artifact

TestRail upload
"""""""""""""""

> To store result we use TestRail. TestRail is a declarative engine
> so before upload result you need to create all test Suites, Plans, Cases,
> Runs, ... you want to.

Reviewer 2: ```suggestion
To store results, we use TestRail which is a declarative engine.
It means that all test suites, plans, cases, runs, etc. must be declared,
before proceeding to the results upload.
```

.. warning::

  TestRail upload is only done for Post-merge as we do not want to

>   store each and every test result coming from branch with some work not
>   completely finished, reviewed or approved.

Reviewer 2: ```suggestion
  store each and every test result coming from branches with on-going work.
```

>   completely finished, reviewed or approved.

Reviewer: I agree with this point, but it makes me think: maybe we should also run the pre-merge tests during post-merge? So we still gather results and stats for these tests on merged code?

Author: I agree, we may want to trigger pre-merge during post-merge, but anyway we do not want to upload result during pre-merge, it's during post-merge always (just post-merge may trigger pre-merge)

  Do not follow this section if it's not a Post-merge test stage.

The file ``eve/testrail_description_file.yaml`` contains all the TestRail
object declarations, that will be created automatically during Post-merge
stage execution.

> It's a simple YAML file with some description used to describe the object
> in TestRail UI.

Reviewer 2: ```suggestion
It's a YAML file used by TestRail UI to describe the objects.
```

> Example:

Reviewer: Is this an exhaustive example? Meaning: can we build more complex stuff than this "suite/section/subsection" hierarchy (e.g., one more sub-subsection level)?

Author: Hmm True we could have sub sub ... sections. Maybe I will change the example to have one level more.

Edit: Or no the example will be too complex, I will add a comment here

Continuous Testing
==================

Add a new test in the continuous integration system
---------------------------------------------------

A test as continuous integration system level is an End-to-End test that need a
dedicated environment (with one or several machines (virtual or container)).

> A test that only check a specific feature of a classic MetalK8s deployment

Reviewer 3: s/check/checks

should be part of PyTest BDD and not integrated as a dedicated stage in
continuous integration system.

How to choose between Pre-merge and Post-merge
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> Choice really depend on the goal of this test.

Reviewer 3: `The choice really depends on the goals of this test `??

> As high level view:

Reviewer 3: s/As high level view/As a high-level view

Pre-merge:

> - Test is not too long (less than 30 minutes).

Reviewer 3: `Test is usually not long and could last less than 30 minutes` or
`Pre-merge test are usually not long and could last less than 30 minutes`
??

> - Test essential feature of the product (installation, expansion, building,

Reviewer 3: s/feature/features

  ...).

Post-merge:

- Test last longer (more than 30 minutes).
- Test "non-essential" feature of the product (upgrade, downgrade, solutions,
  ...).

How to add a stage in continuous integration system
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> Continuous integration system is controled by the ``eve/main.yml`` YAML file.

Reviewer 3: s/controled/controlled

A stage is defined by a worker and a list of steps. Each stage should be in

> ``stages`` section and trigger by ``pre-merge`` or ``post-merge``.

Reviewer 3: Each stage should be in the stages section and is triggered by pre-merge or post-merge

> To know the different kind of worker available, all the builtin steps, how to

Reviewer 3: s/worker/workers

trigger a stage, ... please refer to the eve documentation.

A test stage in MetalK8s context
""""""""""""""""""""""""""""""""

In MetalK8s context each test stage (eve stage that represent a full test)
should generate a status file containing the result of the test, either a
success or a failure, and a JUnit file containing the result of the test
and information about this test.

> To generate these file each stage need some information to generate this

Reviewer 3: Each stage needs the following information to generate the JUnit file: or

To generate the JUnite file, each stage needs the following information

JUnit file:

- The name of the Test Suite this test stage is part of.
- Section path to regroup test in a Test Suite if needed (optional)
- A test name

Before executing all the steps of the test we first generate a failed
result and at the end of the test we generate a success result. So that the
failed result get overrided by the success one if everything goes well.

At the very end this final status should be uploaded no matter if the test went

> well.

Reviewer 3: At the very end, the final status of a test (here pick between: should or must) be uploaded no matter the outcome of the test.

      junit_info: &_my_stage_junit_info
        TEST_SUITE: my test suite
        CLASS_NAME: my section.my sub section
        TEST_NAME: my test
    worker:
      # ...
      # Worker informations
      # ...
    steps:
      - Git: *git_pull
      - ShellCommand: # Generate a failed final status
          <<: *add_final_status_artifact_failed
          env:
            <<: *_env_final_status_artifact_failed
            <<: *_my_stage_junit_info
            STEP_NAME: my-stage
       # ...
       # All test steps should be here !
       # ...
       - ShellCommand: # Generate a success final status
         <<: *add_final_status_artifact_success
         env:
           <<: *_env_final_status_artifact_success
           <<: *_my_stage_junit_info
           STEP_NAME: my-stage
       - Upload: *upload_final_status_artifact

TestRail upload
"""""""""""""""

To store result we use TestRail. TestRail is a declarative engine
so before upload result you need to create all test Suites, Plans, Cases,
Runs, ... you want to.

.. warning::

  TestRail upload is only done for Post-merge as we do not want to
  store each and every test result coming from branch with some work not

>   completely finished, reviewed, approved.

Reviewer 3: reviewed or approved

  Do not follow this section if it's not a Post-merge test stage.

> The file ``eve/testrail_description_file.yaml`` contain all the TestRail

Reviewer 3: s/contain/contains

           <<: *_my_stage_junit_info
           STEP_NAME: my-stage
       - Upload: *upload_final_status_artifact

TestRail upload
"""""""""""""""

To store results, we use TestRail which is a declarative engine.
It means that all test suites, plans, cases, runs, etc. must be declared,
before proceeding to the results upload.

.. warning::

  TestRail upload is only done for Post-merge as we do not want to
  store each and every test result coming from branches with on-going work.

  Do not follow this section if it's not a Post-merge test stage.

The file ``eve/testrail_description_file.yaml`` contains all the TestRail
object declarations, that will be created automatically during Post-merge
stage execution.

It's a YAML file used by TestRail UI to describe the objects.

Example:

.. code-block:: yaml

   My Test Suite:
     description: >-
       My first test suite description
     section:
       My Section:
         description: >-
           My first section description
         sub_sections:
           My sub section:
             description: >-
               My first sub secttion description
             cases:
               My test: {}

>              # sub_sections:  <-- You can have as much subsection level as you
>              #                    want

Reviewer 2: the personal pronoun "you" should be avoided in documentation:
```suggestion
             # sub_sections:  <-- subsections can be nested as deep as needed
```
